test/distributed/fsdp/test_fsdp_use_orig_params.py

```diff
@@ -0,0 +1,875 @@
+
+    class Model(nn.Module):
+        def __init__(self):
+            super().__init__()
+            torch.manual_seed(42)
+            self.lin1 = nn.Linear(5, 5, bias=True)
+            self.lin2 = nn.Linear(5, 7, bias=True)
+
+        def forward(self, x: torch.Tensor) -> torch.Tensor:
+            z = self.lin1(x)
+            z = nn.functional.relu(z)
+            z = self.lin2(z)
+            return z
+
+        def get_input(self, device: torch.device) -> Tuple[torch.Tensor, ...]:
+            return (torch.randn((2, 5)).to(device),)
+
+        def get_loss(self, inp, out):
+            return out.sum()
+
+    @property
+    def world_size(self):
+        # Force a world size of 2 since the tests hard code to the FSDP
+        # sharding strategy
+        return 2
+
+    def _check_param_parity(self, ddp_model: DDP, fsdp_model: FSDP):
+        with FSDP.summon_full_params(fsdp_model):
+            for (n1, p1), (n2, p2) in zip(
+                ddp_model.module.named_parameters(),
+                fsdp_model.named_parameters(),
+            ):
+                self.assertEqual(n1, n2)
+                torch.testing.assert_close(p1, p2)
+
+    @skip_if_lt_x_gpu(2)
+    def test_param_writeback(self):
+        """Tests that changes to the original parameters are written back."""
+        self.run_subtests(
+            {
+                "change_first_weight": [True, False],  # first vs. second `weight`
+                "change_data": [True, False],  # change `.data` vs. variable itself
+            },
+            self._test_param_writeback,
+        )
+
+    def _test_param_writeback(self, change_first_weight: bool, change_data: bool):
+        def transform_param(param: nn.Parameter) -> nn.Parameter:
+            return nn.Parameter(torch.ones_like(param) * 2)
+
+        # Check that the writeback propagates
+        ddp_model = DDP(
+            TestFSDPUseOrigParamsWriteback.Model().cuda(), device_ids=[self.rank]
+        )
+        fsdp_model = FSDP(
+            TestFSDPUseOrigParamsWriteback.Model().cuda(), use_orig_params=True
+        )
+        ddp = ddp_model.module  # for brevity
+        fsdp = fsdp_model.module
+        if change_first_weight:
+            if change_data:
+                ddp.lin1.weight.data = transform_param(ddp.lin1.weight)
+                fsdp.lin1.weight.data = transform_param(fsdp.lin1.weight)
+            else:
+                ddp.lin1.weight = transform_param(ddp.lin1.weight)
+                fsdp.lin1.weight = transform_param(fsdp.lin1.weight)
+        else:
+            if change_data:
+                ddp.lin2.weight.data = transform_param(ddp.lin2.weight)
+                fsdp.lin2.weight.data = transform_param(fsdp.lin2.weight)
+            else:
+                ddp.lin2.weight = transform_param(ddp.lin2.weight)
+                fsdp.lin2.weight = transform_param(fsdp.lin2.weight)
+        self._check_param_parity(ddp_model, fsdp_model)  # triggers a writeback
```

Comment:
where does writeback happen, in summon_full_params? Shall we also check other places where we expect writeback to take place (pre-forward, pre-backward)?